### PR TITLE
feat(rerank): control for columns whose missingness predicts the label

### PR DESCRIPTION
**Pull request by enriquea**
## What this adds

A control for feature columns whose **missingness** predicts the label.

`provenance.py` asks what a predictor was **trained** on. Nothing asked which units it was
**run** on, and the two are independent.

## Why

Measured in a 1,362-gene CHD cohort: the bare flag *"was EVE computed for this gene"* scores
**AUC 0.716** against a ClinGen/GenCC-derived label — above the entire nine-feature gnomAD
constraint axis (0.693). EVE is computed from deep alignments for a curated subset of
proteins; that subset is enriched for well-studied genes; well-studied genes are disease
genes. Gradient-boosted trees learn a default branch direction for NaN, so *"not computed"*
reaches the model as a usable feature.

EVE is unsupervised on alignments, so it **passes every circularity check correctly**. The
existing control cannot see this.

Consequence when the control is applied: the `missense_severity` axis's entire add-one gain
over constraint in CHD (+0.045) goes to **−0.001**, and removing the two offending columns
costs 30% of the model's AUPRC. In three other diseases the same axis is unaffected — so
this is not a blanket correction, it is a per-configuration one.

## Design notes worth reviewing

**A coverage floor is the wrong instrument, and this deliberately isn't one.** On a
19,448-gene spine GTEx eQTL is *sparser* than EVE (21% vs 15% coverage) and completely inert
(presence AUC 0.47–0.51) while EVE reaches 0.64–0.78. A 90% floor would delete four whole
axes for the wrong reason. Sparsity is not the defect; sparsity **correlated with the label**
is — which makes this, exactly like circularity, a property of the feature/**label** pair.
The same column scores 0.78 against epilepsy-DEE and 0.64 against CHD-NDD.

**Statistical selection cannot find it; it rewards it.** A univariate `|AUC−0.5|` filter
ranks EVE highly *because* its missingness tracks the label. That is `provenance.py`'s own
argument transposed, and it is why the control chains **in front of** within-axis selection
rather than after it.

**Two conditions are required to bar a column**, because either alone fails at one end of the
sample-size range: an FDR-significant presence/label association (effect size alone flags
noise in a small cohort) **and** a minimum effect (significance alone flags a 1.5-point
coverage difference at n=200k). The 0.05 effect floor is anchored on measurement — the
reference axes in that cohort span presence AUC 0.503–0.530 and must never be barred, or the
control deletes the baseline it exists to protect.

## Changes

- `algorithms/rerank/leakage.py` — `LeakagePolicy`, `LeakageStat`, `LeakageReport`,
  `presence_leakage`, `resolve_leakage`, `leakage_selector`. Reuses `selection.py`'s
  `univariate_auc` and Bamber-bound `_null_se` so the two filters agree on what a null is.
- `engine.py` — `_compose_selector` chains leakage ahead of within-axis selection. Returns
  `None` when neither control is set rather than an identity wrapper, because
  `ReRanker.score` takes a different path for `selector=None`. Works with `selection=None`:
  the two controls are independent.
- `config.py` — `Config.leakage`, defaulting to `None` so **existing runs are byte-identical**.
  Type-checked in `__post_init__`; a bare float would leave the control silently off while
  the caller believed it was on.

## Tests

25 new, TDD throughout (each watched fail first). Covers both leak directions, the
small-effect-at-large-n and large-effect-at-small-n boundaries, fully-observed and
fully-missing columns, single-class labels, FDR across the supplied columns, per-fold
nesting, and column-order preservation. **Full rerank suite: 133 passed.** flake8 clean.

New files are black-formatted. `engine.py` was hand-matched to black style on the new lines
only — the file has pre-existing formatting drift, and reformatting it wholesale would bury
this change in churn.

## Deliberately not included

Dropping is the bluntest remedy. Restricting the universe to where a column is measured, or
admitting the presence indicator as an explicit feature and requiring the score to beat it,
are both better and belong in separate changes. The latter separates *"the score is
informative"* from *"having the score is informative"*.

## Summary by CodeRabbit

* **New Features**
  * Added optional presence-leakage detection for reranking workflows.
  * Automatically identifies feature columns whose missingness is associated with labels and excludes them from scoring.
  * Added configurable significance and effect-size thresholds.
  * Leakage controls can be used independently or alongside existing feature selection.

* **Validation**
  * Added coverage for detection accuracy, edge cases, threshold handling, selector ordering, and invalid configuration values.

### CODE REVIEW

**Comment by enriquea** [comment]
@coderabbitai review

File: hvantk/algorithms/rerank/engine.py [hunk 1]
```diff
@@ -135,17 +168,30 @@ def rerank(config, _allowed_columns=None, _arm="all", _n_conflicted=0,
...
+        selector = _compose_selector(
+            _axis_selector(config.selection, groups), leakage_policy
+        )
+        recording = _compose_selector(
+            _axis_selector(config.selection, groups, frequency), leakage_policy
+        )
         scores = reranker.score(df, feat_cols, y, selector=recording)
         summary = _selection_summary(
             config, df, y, groups, scores, frequency, _arm, _n_conflicted, _n_unknown
         )
```
[the remark this answers is not in the document]
Comment (enriquea) [reply hvantk/algorithms/rerank/engine.py:223]: Verified and fixed in a8c19e35 — this was a real hole. `_selection_summary` called `select_axis` on the unfiltered `groups`, so a leaking column could land in `global_features` and read as endorsed while no fold ever used it.

Fixed with `_leakage_filtered_groups`, resolved **once over the union of columns** rather than per axis: filtering axis by axis would let the FDR correction see only a narrow axis's columns and hold it to a laxer bar than the nested path applies. An axis left with nothing is dropped rather than passed on empty — a zero-column axis reaching `select_axis` reads as measured-and-contributed-nothing, which is a different statement from barred. Returns the same `groups` object when no policy is set, so existing summaries stay byte-identical.

Three regression tests added: leaking column absent from the filtered groups, emptied axis dropped, identity without a policy.

File: hvantk/algorithms/rerank/leakage.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    No score VALUES are used -- only whether each cell is observed.
+    """
+    from scipy.stats import norm
+
+    if not 0.0 < q <= 1.0:
+        raise ValueError(f"q must be in (0, 1]; got {q}")
+
+    y = np.asarray(y)
+    min_effect = max(0.0, float(min_auc) - 0.5)
+    stats: dict[str, LeakageStat] = {}
```
[the remark this answers is not in the document]
Comment (enriquea) [reply hvantk/algorithms/rerank/leakage.py:125]: Fixed in a8c19e35. `max(0.0, nan)` returns `0.0`, so a NaN `min_auc` silently set `min_effect` to zero — which does not merely disable the floor, it bars **every** FDR-significant column with no effect-size protection at all. That is the exact over-correction the floor exists to prevent, and it would be invisible in the output. `min_auc` is now rejected unless finite and in [0, 1], and `min_effect` is computed from the coerced value.

File: hvantk/algorithms/rerank/config.py [hunk 3]
```diff
@@ -136,6 +148,19 @@ def __post_init__(self):
...
+            # Checked rather than duck-typed: the engine reads `.q` and `.min_auc`, so a
+            # bare float or dict here would raise deep inside a per-fold selector, or worse,
+            # be swallowed and leave the control silently off while the caller believes it
+            # is on. A disabled safety control that looks enabled is the failure to avoid.
+            if not isinstance(self.leakage, LeakagePolicy):
+                raise TypeError(
+                    f"Config.leakage must be a LeakagePolicy or None; got "
+                    f"{type(self.leakage).__name__}. To use defaults, pass "
+                    f"LeakagePolicy()."
+                )
```
[the remark this answers is not in the document]
Comment (enriquea) [reply hvantk/algorithms/rerank/config.py:164]: Fixed in a8c19e35. Both `q` and `min_auc` are now validated in `Config.__post_init__`, for the same reason the `isinstance` check was already there: without it the failure surfaces deep inside a per-fold selector, after the matrix is assembled and scoring has begun, rather than at the point the misconfiguration was made. Two parametrised tests cover each field.

File: hvantk/algorithms/rerank/leakage.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+
+def presence_leakage(
+    X, y, columns, q: float = 0.10, min_auc: float = DEFAULT_MIN_AUC
+) -> dict[str, LeakageStat]:
+    """Score every column by how well its presence indicator alone predicts ``y``.
+
+    No score VALUES are used -- only whether each cell is observed.
+    """
+    from scipy.stats import norm
```
[the remark this answers is not in the document]
Comment (enriquea) [reply hvantk/algorithms/rerank/leakage.py:112]: Declared in a8c19e35 — but not as base, and with two corrections to the finding.

**It pre-dates this PR.** `selection.py` has imported `scipy.stats` since #231 (lines 56 and 83). This adds a third importer to a subsystem that already had two undeclared ones; it is not a dependency this branch introduces.

**The failure described is not reachable.** `scipy>=1.8.0` is a hard `install_requires` of scikit-learn (verified against the installed metadata), and the rerank path imports sklearn well before it touches scipy — `Config`, `ReRanker` and `Evaluator` all do. A bare install cannot enable `config.leakage` at all, because it cannot construct the engine. So there is no install in which leakage fails on a missing scipy but would otherwise have worked.

**Base was the wrong placement.** rerank sits behind the optional `scikit-learn` extra (`ml`, `psroc`, `ancestry`); it is not part of a base install. Putting scipy in `[tool.poetry.dependencies]` unconditionally would burden every user who never touches rerank.

Declared it anyway, as `optional = true` and added to the three extras that already carry scikit-learn, because relying on a transitive dependency is how this silently breaks later — the same reasoning already written above `sorted-nearest` in this file. Filed as hygiene rather than as a fix for a live failure.

**Comment by enriquea** [comment]
@coderabbitai review
